serde_codegen/src/ser.rs

```diff
@@ -226,9 +227,10 @@ fn serialize_tuple_struct(
 
     let type_name = name_expr(builder, item_attrs.name());
     let len = serialize_stmts.len();
+    let let_mut = mut_if_nonzero(cx, len);
 
     quote_block!(cx, {
-        let mut state = try!(_serializer.serialize_tuple_struct($type_name, $len));
+        let $let_mut state = try!(_serializer.serialize_tuple_struct($type_name, $len));
         $serialize_stmts
```

Comment:
@nox suggests rolling this into `$serialize_stmts` and generating:

```rust
let state = try!(...);
let mut state = state;
```

where the initial call is never mut and `$serialize_stmts` includes "let mut state = state" if necessary.